File: src/draw_summary.py
"""Every number in the paper that depends on how many European draws have been trained.

Usage: python3 src/draw_summary.py

The study trains EUR87 repeatedly with different seeds, so a set of statements in the paper
move as draws accumulate: the yield span, the within-ancestry noise floor and its range, the
cross-ancestry weights term, and the whole of the draw-dependence subsection. This script
computes all of them in one pass from the finished sets, so the manuscript can be updated
from measured values rather than from memory, and rechecked at any time.

It also makes one check explicit rather than leaving it to be noticed. The paper claims the
noise floor of the weights term exceeds the frequency and LD components combined, and says
this does not depend on which control pair or which draw is used. That holds only while the
SMALLEST floor still exceeds the LARGEST frequency plus LD sum. The margin has been narrowing
as draws are added (0.545 against 0.395 at three draws, 0.510 against 0.405 at four), so the
claim is re-tested here and reported as PASS or FAIL.

Conventions match the figures: degenerate genes, where a mixed combination has V = 0 and the
components are undefined, are excluded throughout, exactly as fig10 does.
"""
import itertools
from pathlib import Path

import numpy as np
import pandas as pd
from scipy import stats

ROOT = Path(__file__).resolve().parents[1]
MODELS = ROOT / "results" / "models"
DEC = ROOT / "results" / "decomposition"
YRI = "YRI87"


def complete_draws():
    out = []
    for d in sorted(MODELS.glob("EUR87_r*")):
        if len(list(d.glob("chr*.summary.tsv"))) == 22:
            out.append(d.name)
    return out


def yield_row(set_name):
    d = pd.concat([pd.read_csv(f, sep="\t") for f in sorted((MODELS / set_name).glob("chr*.summary.tsv"))],
                  ignore_index=True)
    cv = d[(d.cv_r2 > 0.01) & (d.cv_pval < 0.05)]
    usable = cv[cv.n_model > 0]
    return {"set": set_name, "attempted": len(d), "cleared_cv": len(cv), "usable": len(usable),
            "empty_final": len(cv) - len(usable), "median_r2": float(usable.cv_r2.median())}


def load(path):
    d = pd.read_csv(path, sep="\t")
    return d[~d["degenerate"]]


def shares(path):
    d = load(path)
    t = d[["phi_w", "phi_D", "phi_R"]].abs()
    o = t.div(t.sum(axis=1), axis=0)
    o.columns = ["s_w", "s_D", "s_R"]
    o["gene"] = d.gene.values
    o["dom"] = t.values.argmax(axis=1)
    return o


draws = complete_draws()
print(f"{len(draws)} complete European draws: {', '.join(draws)}\n")

print("=" * 78)
print("YIELD  (paper: the draws span X to Y usable, and A to B clear cross-validation)")
print("=" * 78)
rows = [yield_row(s) for s in draws + [YRI]]
y = pd.DataFrame(rows)
print(y.to_string(index=False))
eur = y[y.set.str.startswith("EUR87")]
yri = y[y.set == YRI].iloc[0]
print(f"\n  usable span across draws: {eur.usable.min():,} to {eur.usable.max():,} "
      f"(width {eur.usable.max() - eur.usable.min()})")
print(f"  Yoruba usable: {yri.usable:,}  -> inside the span: "
      f"{eur.usable.min() <= yri.usable <= eur.usable.max()}")
print(f"  cleared cross-validation, all sets: {min(y.cleared_cv):,} to {max(y.cleared_cv):,}")
frac = [(r.empty_final / r.cleared_cv) for r in y.itertuples()]
print(f"  empty final fit as a share of cleared: {min(frac):.0%} to {max(frac):.0%}")

print("\n" + "=" * 78)
print("NOISE FLOOR  (paper: the floor lies between X and Y)")
print("=" * 78)
floors = []
for a, b in itertools.combinations(draws, 2):
    p = DEC / f"{a}_vs_{b}.tsv"
    if not p.exists():
        print(f"  {a} vs {b}: MISSING, not yet computed")
        continue
    d = load(p)
    floors.append((f"{a}_vs_{b}", len(d), float(d.phi_w.abs().median())))
for name, n, v in floors:
    print(f"  {name:28s} n={n:4d}  median |phi_w| = {v:.3f}")
fv = [v for _, _, v in floors]
print(f"\n  {len(floors)} of {len(draws) * (len(draws) - 1) // 2} possible pairs present")
print(f"  floor range {min(fv):.3f} to {max(fv):.3f}, median {float(np.median(fv)):.3f}")

print("\n" + "=" * 78)
print("CROSS-ANCESTRY COMPONENTS  (paper: weights term lies between X and Y)")
print("=" * 78)
cross = []
for s in draws:
    p = DEC / f"{s}_vs_{YRI}.tsv"
    if not p.exists():
        print(f"  {s}: MISSING")
        continue
    d = load(p)
    w, dd, r = (float(d[c].abs().median()) for c in ("phi_w", "phi_D", "phi_R"))
    cross.append((s, len(d), w, dd, r))
    print(f"  {s:12s} n={len(d):4d}  |phi_w|={w:.3f}  |phi_D|={dd:.3f}  |phi_R|={r:.3f}  D+R={dd + r:.3f}")
cw = [w for _, _, w, _, _ in cross]
sums = [dd + r for _, _, _, dd, r in cross]
print(f"\n  weights term across draws: {min(cw):.3f} to {max(cw):.3f}")
print(f"  frequency plus LD sum:     {min(sums):.3f} to {max(sums):.3f}")

print("\n" + "=" * 78)
print("THE MARGIN CLAIM  (paper: the floor exceeds frequency and LD combined,")
print("                   whichever control pair and whichever draw is used)")
print("=" * 78)
lo_floor, hi_sum = min(fv), max(sums)
print(f"  smallest floor      {lo_floor:.3f}")
print(f"  largest D+R sum     {hi_sum:.3f}")
print(f"  margin              {lo_floor - hi_sum:+.3f}")
print(f"  VERDICT: {'PASS, the claim holds for every pairing' if lo_floor > hi_sum else 'FAIL, the sentence must be rewritten'}")

print("\n" + "=" * 78)
print("DRAW DEPENDENCE  (paper: pooling all pairs gives N comparisons over G genes,")
print("                  the largest component changes for X percent)")
print("=" * 78)
per = {s: shares(DEC / f"{s}_vs_{YRI}.tsv") for s in draws if (DEC / f"{s}_vs_{YRI}.tsv").exists()}
merged = [per[a].merge(per[b], on="gene", suffixes=("_1", "_2")) for a, b in itertools.combinations(per, 2)]
m = pd.concat(merged, ignore_index=True)
M = np.zeros((3, 3), int)
for i in range(3):
    for j in range(3):
        M[i, j] = int(((m.dom_1 == i) & (m.dom_2 == j)).sum())
agree = float((m.dom_1 == m.dom_2).mean())
print(f"  {len(merged)} pairs, {len(m):,} pooled comparisons, {m.gene.nunique()} distinct genes")
print(f"  agreement {agree:.1%}, so the largest component changes for {1 - agree:.1%}")
print(f"  matrix diagonal: weights {M[0, 0]}, frequency {M[1, 1]}, LD {M[2, 2]}; switched {int(M.sum() - np.trace(M))}")
print(f"  weights-share Spearman {stats.spearmanr(m.s_w_1, m.s_w_2)[0]:.2f}, "
      f"median absolute change {float((m.s_w_1 - m.s_w_2).abs().median()):.2f}")
print("  per-pair agreement:")
for (a, b), mm in zip(itertools.combinations(per, 2), merged):
    print(f"    {a} vs {b}: {len(mm)} genes, {(mm.dom_1 == mm.dom_2).mean():.1%}")

print("\n  median share of the total, per draw (Figure 13 panel c):")
for s in per:
    d = per[s]
    print(f"    {s:12s} weights {d.s_w.median():.3f}  frequency {d.s_D.median():.3f}  LD {d.s_R.median():.3f}")
sw = [per[s].s_w.median() for s in per]
sd = [per[s].s_D.median() for s in per]
sr = [per[s].s_R.median() for s in per]
print(f"    ranges: weights {min(sw):.3f} to {max(sw):.3f}, "
      f"frequency {min(sd):.3f} to {max(sd):.3f}, LD {min(sr):.3f} to {max(sr):.3f}")

File: src/draw_summary.R
## Every number in the paper that depends on how many European draws have been trained.
## Also re-tests the margin claim: the smallest noise floor must still exceed the
## largest frequency plus LD sum, and it reports PASS or FAIL.
##
## Usage: Rscript src/draw_summary.R

suppressPackageStartupMessages(library(data.table))

find_root <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grep("^--file=", a)])
  d <- if (length(f)) dirname(normalizePath(f[1L])) else getwd()
  normalizePath(file.path(d, ".."), mustWork = FALSE)
}
ROOT   <- find_root()
MODELS <- file.path(ROOT, "results", "models")
DEC    <- file.path(ROOT, "results", "decomposition")
YRI    <- "YRI87"

complete_draws <- function() {
  dirs <- list.dirs(MODELS, recursive = FALSE)
  dirs <- dirs[grepl("EUR87_r", basename(dirs))]
  keep <- vapply(dirs, function(d) length(list.files(d, pattern = "^chr.*\\.summary\\.tsv$")) == 22L, logical(1))
  sort(basename(dirs[keep]))
}

yield_row <- function(set_name) {
  files <- sort(list.files(file.path(MODELS, set_name), pattern = "^chr.*\\.summary\\.tsv$", full.names = TRUE))
  d  <- rbindlist(lapply(files, fread, sep = "\t"), fill = TRUE)
  cv <- d[cv_r2 > 0.01 & cv_pval < 0.05]
  usable <- cv[n_model > 0]
  data.table(set = set_name, attempted = nrow(d), cleared_cv = nrow(cv), usable = nrow(usable),
             empty_final = nrow(cv) - nrow(usable), median_r2 = median(usable$cv_r2))
}

load_dec <- function(path) {
  d <- fread(path, sep = "\t")
  d[degenerate == FALSE | degenerate == "False"]
}

shares <- function(path) {
  d <- load_dec(path)
  t <- abs(d[, .(phi_w, phi_D, phi_R)])
  s <- rowSums(t)
  o <- as.data.table(t / s)
  setnames(o, c("s_w", "s_D", "s_R"))
  o[, gene := d$gene]
  o[, dom := max.col(t, ties.method = "first") - 1L]
  o[]
}

draws <- complete_draws()
cat(sprintf("%d complete European draws: %s\n\n", length(draws), paste(draws, collapse = ", ")))

cat(strrep("=", 78), "\n", sep = "")
cat("YIELD  (paper: the draws span X to Y usable, and A to B clear cross-validation)\n")
cat(strrep("=", 78), "\n", sep = "")
y <- rbindlist(lapply(c(draws, YRI), yield_row))
print(y)
eur <- y[grepl("^EUR87", set)]
yri <- y[set == YRI]
cat(sprintf("\n  usable span across draws: %s to %s (width %d)\n",
            format(min(eur$usable), big.mark = ","), format(max(eur$usable), big.mark = ","),
            max(eur$usable) - min(eur$usable)))
cat(sprintf("  Yoruba usable: %s  -> inside the span: %s\n",
            format(yri$usable, big.mark = ","),
            min(eur$usable) <= yri$usable && yri$usable <= max(eur$usable)))
cat(sprintf("  cleared cross-validation, all sets: %s to %s\n",
            format(min(y$cleared_cv), big.mark = ","), format(max(y$cleared_cv), big.mark = ",")))
frac <- y$empty_final / y$cleared_cv
cat(sprintf("  empty final fit as a share of cleared: %.0f%% to %.0f%%\n", 100 * min(frac), 100 * max(frac)))

cat("\n", strrep("=", 78), "\n", sep = "")
cat("NOISE FLOOR  (paper: the floor lies between X and Y)\n")
cat(strrep("=", 78), "\n", sep = "")
floors <- list()
if (length(draws) >= 2L) {
  cb <- combn(draws, 2L)
  for (i in seq_len(ncol(cb))) {
    a <- cb[1L, i]; b <- cb[2L, i]
    p <- file.path(DEC, sprintf("%s_vs_%s.tsv", a, b))
    if (!file.exists(p)) { cat(sprintf("  %s vs %s: MISSING, not yet computed\n", a, b)); next }
    d <- load_dec(p)
    floors[[length(floors) + 1L]] <- list(name = sprintf("%s_vs_%s", a, b), n = nrow(d),
                                          v = median(abs(d$phi_w)))
  }
}
for (f in floors) cat(sprintf("  %-28s n=%4d  median |phi_w| = %.3f\n", f$name, f$n, f$v))
fv <- vapply(floors, function(f) f$v, numeric(1L))
cat(sprintf("\n  %d of %d possible pairs present\n", length(floors), choose(length(draws), 2L)))
cat(sprintf("  floor range %.3f to %.3f, median %.3f\n", min(fv), max(fv), median(fv)))

cat("\n", strrep("=", 78), "\n", sep = "")
cat("CROSS-ANCESTRY COMPONENTS  (paper: weights term lies between X and Y)\n")
cat(strrep("=", 78), "\n", sep = "")
cw <- numeric(0); sums <- numeric(0)
for (s in draws) {
  p <- file.path(DEC, sprintf("%s_vs_%s.tsv", s, YRI))
  if (!file.exists(p)) { cat(sprintf("  %s: MISSING\n", s)); next }
  d  <- load_dec(p)
  w  <- median(abs(d$phi_w)); dd <- median(abs(d$phi_D)); r <- median(abs(d$phi_R))
  cw <- c(cw, w); sums <- c(sums, dd + r)
  cat(sprintf("  %-12s n=%4d  |phi_w|=%.3f  |phi_D|=%.3f  |phi_R|=%.3f  D+R=%.3f\n",
              s, nrow(d), w, dd, r, dd + r))
}
cat(sprintf("\n  weights term across draws: %.3f to %.3f\n", min(cw), max(cw)))
cat(sprintf("  frequency plus LD sum:     %.3f to %.3f\n", min(sums), max(sums)))

cat("\n", strrep("=", 78), "\n", sep = "")
cat("THE MARGIN CLAIM  (paper: the floor exceeds frequency and LD combined,\n")
cat("                   whichever control pair and whichever draw is used)\n")
cat(strrep("=", 78), "\n", sep = "")
lo_floor <- min(fv); hi_sum <- max(sums)
cat(sprintf("  smallest floor      %.3f\n", lo_floor))
cat(sprintf("  largest D+R sum     %.3f\n", hi_sum))
cat(sprintf("  margin              %+.3f\n", lo_floor - hi_sum))
cat(sprintf("  VERDICT: %s\n", if (lo_floor > hi_sum) "PASS, the claim holds for every pairing"
                               else "FAIL, the sentence must be rewritten"))

cat("\n", strrep("=", 78), "\n", sep = "")
cat("DRAW DEPENDENCE  (paper: pooling all pairs gives N comparisons over G genes,\n")
cat("                  the largest component changes for X percent)\n")
cat(strrep("=", 78), "\n", sep = "")
per <- list()
for (s in draws) {
  p <- file.path(DEC, sprintf("%s_vs_%s.tsv", s, YRI))
  if (file.exists(p)) per[[s]] <- shares(p)
}
nms <- names(per)
merged <- list(); pair_names <- list()
if (length(nms) >= 2L) {
  cb <- combn(nms, 2L)
  for (i in seq_len(ncol(cb))) {
    mm <- merge(per[[cb[1L, i]]], per[[cb[2L, i]]], by = "gene", suffixes = c("_1", "_2"))
    merged[[length(merged) + 1L]] <- mm
    pair_names[[length(pair_names) + 1L]] <- paste(cb[1L, i], "vs", cb[2L, i])
  }
}
m <- rbindlist(merged)
M <- matrix(0L, 3L, 3L)
for (i in 0:2) for (j in 0:2) M[i + 1L, j + 1L] <- sum(m$dom_1 == i & m$dom_2 == j)
agree <- mean(m$dom_1 == m$dom_2)
cat(sprintf("  %d pairs, %s pooled comparisons, %d distinct genes\n",
            length(merged), format(nrow(m), big.mark = ","), uniqueN(m$gene)))
cat(sprintf("  agreement %.1f%%, so the largest component changes for %.1f%%\n", 100 * agree, 100 * (1 - agree)))
cat(sprintf("  matrix diagonal: weights %d, frequency %d, LD %d; switched %d\n",
            M[1, 1], M[2, 2], M[3, 3], sum(M) - sum(diag(M))))
cat(sprintf("  weights-share Spearman %.2f, median absolute change %.2f\n",
            cor(m$s_w_1, m$s_w_2, method = "spearman"), median(abs(m$s_w_1 - m$s_w_2))))
cat("  per-pair agreement:\n")
for (i in seq_along(merged))
  cat(sprintf("    %s: %d genes, %.1f%%\n", pair_names[[i]], nrow(merged[[i]]),
              100 * mean(merged[[i]]$dom_1 == merged[[i]]$dom_2)))

cat("\n  median share of the total, per draw (Figure 13 panel c):\n")
for (s in nms) {
  d <- per[[s]]
  cat(sprintf("    %-12s weights %.3f  frequency %.3f  LD %.3f\n",
              s, median(d$s_w), median(d$s_D), median(d$s_R)))
}
sw <- vapply(per, function(d) median(d$s_w), numeric(1L))
sd_ <- vapply(per, function(d) median(d$s_D), numeric(1L))
sr <- vapply(per, function(d) median(d$s_R), numeric(1L))
cat(sprintf("    ranges: weights %.3f to %.3f, frequency %.3f to %.3f, LD %.3f to %.3f\n",
            min(sw), max(sw), min(sd_), max(sd_), min(sr), max(sr)))
